src/Mono.Android/Android.Runtime/AndroidRuntime.cs

```diff
@@ -283,37 +283,54 @@ static bool FastRegisterNativeMembers (JniType nativeClass, Type type, string me
 		}
 
 		class MagicRegistrationMap {
-			static Dictionary<string, int> typesMap;
-
-			static void Prefill ()
-			{
-				// fill code added by the linker
-			}
+			// should stay in sync with MonoDroidMarkStep.GetStringHashCode
+			static int GetStringHashCode (string str)
+	                {
+	                        int hash1 = 5381;
+	                        int hash2 = hash1;
+
+	                        unsafe {
+	                                fixed (char *src = str) {
+	                                        int c;
+	                                        char *s = src;
+	                                        while ((c = s[0]) != 0) {
```

Comment:
This doesn't look safe if `str` is null.

It also appears to require that `str` be `null`-terminated to work with strings with `str.Length==0`.  I'm not sure if this is a portable assumption; it's true for mono, but is it true for .NET? Will it be true in the future?

I'd kinda prefer a `for` loop and removal of the conditional:

```csharp
int i;
int len = str.Length;
for (i = 0; (i+1) < len; i += 2) {
    hash1 = ((hash1 << 5) + hash1) ^ src [i];
    hash2 = ((hash2 << 5) + hash2) ^ src [i+1];
}
if (i < len) {
    hash1 = ((hash1 << 5) + hash1) ^ src [i];
}
```

Yes, this repeats the expression for `hash1`, but it removes a conditional from the primary loop.  It would be interesting to see how the performance compares.